### PR TITLE
DEV: Adds RRF algorithm and API for adding results to search

### PR DESCRIPTION
**Pull request by keegangeorge**
**This PR adds an API necessary for usage by Discourse AI's semantic search feature.**

**Specifically it:**
- Adds `@action` method called `addSearchResults()` to `full-page-search` which can by `full-page-search-below-search-header` plugin outlet to add additional search results to the existing initial search results found in the `full-page-search`. This will be used by Discourse AI to add AI semantic search results in parallel with normal search.
- `addSearchResults` are then mixed with initial results using [Reciprocal Ranking Fusion Algorithm](https://plg.uwaterloo.ca/~gvcormac/cormacksigir09-rrf.pdf)
   - Method for handling this reciprocal ranking is added to `lib/search.js` and used inside full page search when handling the `addSearchResults`

See corresponding PR in Discourse AI: https://github.com/discourse/discourse-ai/pull/278

### CODE REVIEW

File: app/assets/javascripts/discourse/app/controllers/full-page-search.js [hunk 1]
```diff
@@ -464,9 +499,14 @@ export default Controller.extend({
     });
   },
 
+  @action
+  recheckSearchResults() {
+    this.notifyPropertyChange("searchResultPosts");
+  },
+
```
Comment (keegangeorge) [inline app/assets/javascripts/discourse/app/controllers/full-page-search.js:506]: Instead of having to do:

```js
     api.addAdditionalSearchResults(AIResults);
     this.args.outletArgs.recheckSearchResults();
```

inside Discourse AI plugin to refresh search results, is there an easy way to automatically trigger `recheckSearchResults` in core whenever the API method is called?

I suppose we can use `getOwner`, `getController` and call the method after the pluginAPI method is called, but it feels like it introduces code smell. Though, I still feel this recheckingSearchResults should be automatically handled by the API rather than having to be invoked in the plugin.

File: app/assets/javascripts/discourse/app/controllers/full-page-search.js [hunk 2]
```diff
@@ -308,6 +315,34 @@ export default Controller.extend({
...
+      // Apply RRF to results from each source:
+      const mixedResults = [
+        ...reciprocalRankingAlgorithm(posts),
+        ...reciprocalRankingAlgorithm(additionalSearchResults),
+      ];
+
+      // Sort results by RRF
+      const sortedResults = sortByReciprocalRanking(mixedResults);
+
+      // Remove duplicate results
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/controllers/full-page-search.js:330]: Since reciprocal ranking only sorts results, it may make sense to remove duplicates before mixing in and sorting the additional items.

File: app/assets/javascripts/discourse/app/lib/plugin-api.js [hunk 3]
```diff
@@ -2559,6 +2562,24 @@ class PluginApi {
...
+
+  /**
+   * Allows you add additional search results to the full page search results.
+   * Search results added through this API will be mixed with initial search results
+   * and ranked using the reciprocal ranking fusion algorithm.
+   *
+   * (See `reciprocalRankingAlgorithm()` in `discourse/app/lib/search.js` for more
+   * details on how reciprocal ranking works)
+   *
+   * Full page must be refreshed using `recheckSearchResults()` to see the new results.
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/lib/plugin-api.js:2574]: This may be a bit confusing, people may think "Full page" here means a reload is needed. 
```suggestion
   * Use `recheckSearchResults()` to trigger a refresh of the component and see the new results.
```

And on second thought, shouldn't this be automatically done?  I.e. `api.addAdditionalSearchResults` should queue a rerender of the results.

File: app/assets/javascripts/discourse/app/lib/plugin-api.js [hunk 4]
```diff
@@ -2559,6 +2562,24 @@ class PluginApi {
...
+   *
+   * Example:
+   *
+   * api.addAdditionalSearchResults(newPostArray);
+   */
+  addAdditionalSearchResults(results) {
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/lib/plugin-api.js:2580]: Using `add` and `Additional` doesn't sound quite right, it's a little repetitive. Maybe there's a better name for this?

File: app/assets/javascripts/discourse/app/lib/plugin-api.js (hunk 3 above)
Comment (keegangeorge) [reply app/assets/javascripts/discourse/app/lib/plugin-api.js:2574]: > And on second thought, shouldn't this be automatically done? I.e. api.addAdditionalSearchResults should queue a rerender of the results.

Yes this is what I would like to have done automatically. Though, I'm unsure how best to achieve this since the action is inside the controller. Please see https://github.com/discourse/discourse/pull/24202#discussion_r1379230590. I'm wondering how to do this without introducing code smell... Using `getController` and calling the action is bad practice?

File: app/assets/javascripts/discourse/app/lib/plugin-api.js (hunk 4 above)
Comment (keegangeorge) [reply app/assets/javascripts/discourse/app/lib/plugin-api.js:2580]: Ah yes I agree. How about one of the these:
- `addSearchResults`
- `addCustomSearchResults`
- `addMoreSearchResults`
- `addToSearchResults`

Or any other suggestion you may have? I'm not particular tied to any of them.

File: app/assets/javascripts/discourse/app/components/search-result-entry.js [hunk 5]
```diff
@@ -6,7 +6,7 @@ import { modKeysPressed } from "discourse/lib/utilities";
 export default Component.extend({
   tagName: "div",
   classNames: ["fps-result"],
-  classNameBindings: ["bulkSelectEnabled"],
+  classNameBindings: ["bulkSelectEnabled", "post.generatedByAI:ai-result"],
```
Comment (xfalcox) [inline app/assets/javascripts/discourse/app/components/search-result-entry.js:9]: Why is this here?

File: app/assets/javascripts/discourse/app/lib/search.js [hunk 6]
```diff
@@ -259,3 +259,40 @@ export function logSearchLinkClick(params) {
...
+
+ * Sorts an array of mixed search results by their reciprocal rank
+ * Before sorting, use reciprocalRankingAlgorithm() to assign a reciprocal rank to each result
+ *
+ * @param {Array} results - an combined array of search results from multiple sources
+ *
+ **/
+export function sortByReciprocalRanking(results) {
+  return results.sort((a, b) => b.reciprocalRank - a.reciprocalRank);
+}
```
Comment (xfalcox) [inline app/assets/javascripts/discourse/app/lib/search.js:298]: Can't we combine those and expose a single API that receives a list of lists and returns a single sorted list ?

File: app/assets/javascripts/discourse/app/lib/plugin-api.js (hunk 4 above)
Comment (pmusaraj) [reply app/assets/javascripts/discourse/app/lib/plugin-api.js:2580]: `addSearchResults` is probably my favourite from the list.

File: app/assets/javascripts/discourse/app/controllers/full-page-search.js (hunk 1 above)
Comment (pmusaraj) [reply app/assets/javascripts/discourse/app/controllers/full-page-search.js:506]: Would it work to do something like: 

```
  @discourseComputed("model.posts", "additionalSearchResults")
  searchResultPosts(posts) {
```

So basically add a way to watch the list of additional search results and recalculate `searchResultPosts` whenever that changes.

Comment (keegangeorge) [reply app/assets/javascripts/discourse/app/controllers/full-page-search.js:506]: This doesn't seem to work. I think it's not working because `additionalSearchResults` is defined outside of the controller so that it can be used along with the exported plugin API function:

https://github.com/discourse/discourse/blob/639b88f07ea96fd76c86592e1d1d7ad001d51e26/app/assets/javascripts/discourse/app/controllers/full-page-search.js#L47

https://github.com/discourse/discourse/blob/639b88f07ea96fd76c86592e1d1d7ad001d51e26/app/assets/javascripts/discourse/app/controllers/full-page-search.js#L57-L59

Comment (keegangeorge) [reply app/assets/javascripts/discourse/app/controllers/full-page-search.js:486]: I'm going to go ahead and just solve this by doing:

```diff
  /**
   * Allows you add additional search results to the full page search results.
   * Search results added through this API will be mixed with initial search results
   * and ranked using the reciprocal ranking fusion algorithm.
   *
   * (See `reciprocallyRankedList()` in `discourse/app/lib/search.js` for more
   * details on how reciprocal ranking works)
   *
   *
   * Example:
   *
   * api.addSearchResults(newPostArray);
   */
  addSearchResults(results) {
    additionalResults(results);
+    this.container.lookup("controller:full-page-search").recheckSearchResults();
  }
```

Not sure if this is the best solution, but it works for now. We can always revisit and update the implementation in a separate PR if we can find a better solution.

File: app/assets/javascripts/discourse/app/lib/plugin-api.js [hunk 7]
```diff
@@ -2559,6 +2562,24 @@ class PluginApi {
...
+   *
+   * (See `reciprocallyRankedList()` in `discourse/app/lib/search.js` for more
+   * details on how reciprocal ranking works)
+   *
+   *
+   * Example:
+   *
+   * api.addSearchResults(newPostArray);
+   */
+  addSearchResults(results) {
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/lib/plugin-api.js:2579]: I'm not sure about this kind of 'push' API. How is the plugin/theme supposed to actually do their ajax request? Or find out what the search term is? They have to connect to a Plugin Outlet on the search page and observe the arguments?

I feel like this API would make more sense as some kind of callback, which can return a promise. Then themes/plugins could do something like:

```js
addSearchResultsCallback(await (term) => {
  const results = await ajax(`my-search-endpoint.json?term=${term}`)
  results.topics;
});
```

File: app/assets/javascripts/discourse/app/controllers/full-page-search.js [hunk 8]
```diff
@@ -464,9 +479,15 @@ export default Controller.extend({
     });
   },
 
+  @action
+  recheckSearchResults() {
+    // Note: Action is called by pluginAPI method `addSearchResults()`
+    this.notifyPropertyChange("searchResultPosts");
+  },
+
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/app/controllers/full-page-search.js:487]: To avoid the need for this, we could:

1. Make `additionalSearchResults` a TrackedArray

2. Convert `searchResultPosts` to a native getter, so that it automatically re-evaluates when any trackable state changes

File: app/assets/javascripts/discourse/app/components/search-result-entry.hbs [hunk 9]
```diff
@@ -78,4 +78,6 @@
       </span>
     {{/if}}
   {{/if}}
-</div>
\ No newline at end of file
+</div>
+
+<PluginOutlet @name="after-search-result-entry" @connectorTagName="div" />
```
Comment (CvX) [inline app/assets/javascripts/discourse/app/components/search-result-entry.hbs:83]: does it need the extra div?

```suggestion
<PluginOutlet @name="after-search-result-entry" />
```

File: app/assets/javascripts/discourse/app/controllers/full-page-search.js [hunk 10]
```diff
@@ -252,15 +254,15 @@ export default Controller.extend({
     return I18n.t("search.result_count", { count, plus, term });
   },
 
-  @observes("model.[posts,categories,tags,users].length")
+  @observes("model.[posts,categories,tags,users].length", "searchResultPosts")
```
Comment (CvX) [inline app/assets/javascripts/discourse/app/controllers/full-page-search.js:257]: I know this was already there but… does this dependent key syntax even work? 😄 cc: @davidtaylorhq 

It should be `{}`?

```js
  @observes("model.{posts,categories,tags,users}.length", "searchResultPosts")
```

(see https://guides.emberjs.com/v3.12.0/object-model/computed-properties/ and https://guides.emberjs.com/v3.12.0/object-model/computed-properties-and-aggregate-data/ for the old docs)

File: app/assets/javascripts/discourse/app/controllers/full-page-search.js [hunk 11]
```diff
@@ -308,6 +310,18 @@ export default Controller.extend({
       : "search-info";
   },
 
+  @discourseComputed("model.posts", "additionalSearchResults")
+  searchResultPosts(posts, additionalSearchResults) {
+    if (additionalSearchResults && additionalSearchResults?.list?.length > 0) {
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/controllers/full-page-search.js:315]:
```suggestion
  @discourseComputed("model.posts", "additionalSearchResults.list")
  searchResultPosts(posts, additionalResultsList) {
    if (additionalResultsList?.length > 0) {
```

Very minor, but I don't think we need both `additionalSearchResults &&` and `additionalSearchResults?` in this clause. In fact, maybe we can take this a step further as per above (the reference to `additionalSearchResults.list` in line 317 below would need to be updated as well).

File: app/assets/javascripts/discourse/app/lib/plugin-api.js [hunk 12]
```diff
@@ -136,7 +136,7 @@ import { modifySelectKit } from "select-kit/mixins/plugin-api";
 // docs/CHANGELOG-JAVASCRIPT-PLUGIN-API.md whenever you change the version
 // using the format described at https://keepachangelog.com/en/1.0.0/.
 
-export const PLUGIN_API_VERSION = "1.15.0";
+export const PLUGIN_API_VERSION = "1.16.0";
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/lib/plugin-api.js:139]: Does `reciprocallyRankedList` need to be exposed in the plugin API? It looks like it isn't necessary, in which case we don't need to bump the version.

File: app/assets/javascripts/discourse/tests/unit/lib/search-test.js [hunk 13]
```diff
@@ -62,4 +63,190 @@ module("Unit | Utility | search", function (hooks) {
...
+      [sourceA, sourceB],
+      ["topic_id", "topic_id"]
+    );
+
+    assert.deepEqual(
+      rankedList,
+      desiredMixedResults,
+      "it correctly ranks the results using the reciprocal ranking algorithm"
+    );
+  });
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/tests/unit/lib/search-test.js:251]: Can you add a test case with 3 or 4 lists? I also wonder how the ranking algorithm handles different data stuctures in lists, maybe add a bit of variety in there as well.

**Review by pmusaraj — APPROVED** [review]
Looks generally good to me, just a couple of comments.

File: app/assets/javascripts/discourse/app/lib/plugin-api.js (hunk 12 above)
Comment (keegangeorge) [reply app/assets/javascripts/discourse/app/lib/plugin-api.js]: Ah yes good catch, thanks!
